Modify pip in appcompatprocessor.sls, update pdftk

digitalsleuth opened the pull request:
Specific pointer in appcompatprocessor.sls to /usr/bin/pip can cause conflict with locally installed pip. Using the salt-based pip uses 'python2 -m pip' which is more preferable for this installation. Also removed the specific git commit as the tool has been updated for additional support for Windows 10.

The pdftk installation from malteworld repo has been removed (unsure if permanent), making it impossible to install pdftk from the repo. Added pdftk-java state to compensate for this, allowing for a direct download of the deb package, then local install with requirements. 

Updated the init.sls, did not remove the repo (in the event pdftk is returned).

These two fixes should resolve the issues in issue 460.

ekristen commented:
Nice. I’ll review soon along with your other changes and test and will report back.

ekristen commented:
This did not work for me. appcompact failed to install. :/

digitalsleuth commented:
What was the error you got? It installed without error here. This is just from a few seconds ago.
salt-call -l debug --local --retcode-passthrough --state-output=mixed state.sls sift.python-packages.appcompatprocessor

local:
  Name: git - Function: pkg.installed - Result: Changed Started: - 17:20:28.768673 Duration: 19687.321 ms
  Name: python - Function: pkg.installed - Result: Clean Started: - 17:20:48.472636 Duration: 1501.338 ms
  Name: python-pip - Function: pkg.installed - Result: Changed Started: - 17:20:49.974461 Duration: 33504.248 ms
  Name: git+https://github.com/mbevilacqua/appcompatprocessor.git - Function: pip.installed - Result: Changed Started: - 17:21:25.552167 Duration: 45072.114 ms

Summary for local
------------
Succeeded: 4 (changed=3)
Failed:    0
------------
Total states run:     4
Total run time:  99.765 s

________________________________
From: Erik Kristensen <notifications@github.com>
Sent: Monday, May 18, 2020 1:18 PM
To: teamdfir/sift-saltstack <sift-saltstack@noreply.github.com>
Cc: Corey Forman <corey@digitalsleuth.ca>; Author <author@noreply.github.com>
Subject: Re: [teamdfir/sift-saltstack] Modify pip in appcompatprocessor.sls, update pdftk (#28)


This did not work for me. appcompact failed to install. :/

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/teamdfir/sift-saltstack/pull/28#issuecomment-630322968>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AO7OHXT3ZD635KM7ZJCNKSTRSFUWVANCNFSM4NDLQMJA>.

ekristen commented:
From what state were you testing this? Clean system with nothing else installed? I think this is part of the problem. Python 2 vs 3 depending on installation order becomes the default and weird things pursue. I just did a full VM build and it was unsuccesful to build appcompat, but the pdftk was successful.

digitalsleuth commented:
I'm using the dockers within the sift-saltstack repo, launching the install using:
`sudo ./.ci/dev.sh appcompat`

digitalsleuth commented:
Just built a brand new minimal Ubuntu 18.04 install VM. Apt update/upgrade -y.
Followed the steps in the readme to add salt-minion, added git. Cloned the repo to /srv/salt, changed appcompatprocessor.sls to the above configured PR, and ran:
`sudo salt-call -l debug --local --file-root=/srv/salt state.apply sift.python-packages.appcompatprocessor`.

This ran successfully with no errors as well.

digitalsleuth commented:
Also just manually changed my default python to 3 using update-alternatives, set to manual. Installed python3-pip, set it to the default pip binary and ran the command again. Still installed without errors, as salt seems to call /usr/bin/python2 -m pip to do the install.

Any chance you can share the error and debug output you received so I can duplicate and help figure out a workaround/fix?

Cheers!

ekristen commented:
```
    amazon-ebs: [INFO    ] {'git+https://github.com/dkovar/analyzeMFT.git@64c71d7c8905a119b7abdf9813e6ef5f11d3ccf1==???': 'Installed'}
    amazon-ebs: [INFO    ] Completed state [git+https://github.com/dkovar/analyzeMFT.git@64c71d7c8905a119b7abdf9813e6ef5f11d3ccf1] at time 14:12:27.968839 (duration_in_ms=2947.15)
    amazon-ebs: [INFO    ] Running state [git+https://github.com/mbevilacqua/appcompatprocessor.git] at time 14:12:27.972010
    amazon-ebs: [INFO    ] Executing state pip.installed for [git+https://github.com/mbevilacqua/appcompatprocessor.git]
    amazon-ebs: [INFO    ] Executing command ['/usr/bin/python3', '-m', 'pip', '--version'] in directory '/home/sansforensics'
    amazon-ebs: [INFO    ] Executing command ['/usr/bin/python3', '-m', 'pip', 'freeze', '--all'] in directory '/home/sansforensics'
    amazon-ebs: [INFO    ] Executing command ['/usr/bin/python3', '-m', 'pip', 'install', '--disable-pip-version-check', 'git+https://github.com/mbevilacqua/appcompatprocessor.git'] in directory '/home/sansforensics'
    amazon-ebs: [ERROR   ] Command '['/usr/bin/python3', '-m', 'pip', 'install', '--disable-pip-version-check', 'git+https://github.com/mbevilacqua/appcompatprocessor.git']' failed with return code: 1
    amazon-ebs: [ERROR   ] stdout: Collecting git+https://github.com/mbevilacqua/appcompatprocessor.git
    amazon-ebs:   Cloning https://github.com/mbevilacqua/appcompatprocessor.git to /tmp/pip-6fnpioh9-build
    amazon-ebs:     Complete output from command python setup.py egg_info:
    amazon-ebs:     Traceback (most recent call last):
    amazon-ebs:       File "<string>", line 1, in <module>
    amazon-ebs:       File "/tmp/pip-6fnpioh9-build/setup.py", line 4, in <module>
    amazon-ebs:         from settings import __version__
    amazon-ebs:       File "/tmp/pip-6fnpioh9-build/settings.py", line 9, in <module>
    amazon-ebs:         from Queue import Empty
    amazon-ebs:     ModuleNotFoundError: No module named 'Queue'
    amazon-ebs:
    amazon-ebs:     ----------------------------------------
    amazon-ebs: [ERROR   ] stderr: The directory '/home/sansforensics/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
    amazon-ebs: The directory '/home/sansforensics/.cache/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
    amazon-ebs: Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-6fnpioh9-build/
    amazon-ebs: [ERROR   ] retcode: 1
    amazon-ebs: [ERROR   ] Failed to install packages: git+https://github.com/mbevilacqua/appcompatprocessor.git. Error: Collecting git+https://github.com/mbevilacqua/appcompatprocessor.git
    amazon-ebs:   Cloning https://github.com/mbevilacqua/appcompatprocessor.git to /tmp/pip-6fnpioh9-build
    amazon-ebs:     Complete output from command python setup.py egg_info:
    amazon-ebs:     Traceback (most recent call last):
    amazon-ebs:       File "<string>", line 1, in <module>
    amazon-ebs:       File "/tmp/pip-6fnpioh9-build/setup.py", line 4, in <module>
    amazon-ebs:         from settings import __version__
    amazon-ebs:       File "/tmp/pip-6fnpioh9-build/settings.py", line 9, in <module>
    amazon-ebs:         from Queue import Empty
    amazon-ebs:     ModuleNotFoundError: No module named 'Queue'
    amazon-ebs:
    amazon-ebs:     ---------------------------------------- The directory '/home/sansforensics/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
    amazon-ebs: The directory '/home/sansforensics/.cache/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
    amazon-ebs: Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-6fnpioh9-build/
```

digitalsleuth commented:
Any chance you can run it in debug? I can see a few major differences with your output than mine. The first is obviously it's using python3, but it's also disabling the pip version check. I'm running mine again and will try to see where it's not shifting from python3 to python2. 

The only other way around this I can think of is to do a cmd.run or cmd.wait to specifically do '/usr/bin/python2 -m pip install ...' We won't be able to point to a specific pip binary anymore because python complains when you do.

Regarding pdftk-java, there's a newer version (3.1.1) which I haven't tested yet, but is in the same repo as in the state attached to this PR. That might be a better option vice the one I added, but I haven't tested it yet. The one in the state is tested.

ekristen commented:
Yeah I can. This is the problem. There's some weird case that I haven't been able to figure out yet where python3 takes precedence. This python2 vs python3 is becoming an issue. We maybe have to simply exclude ALL tools that cannot work in python3 until authors decide to update for python3.

digitalsleuth commented:
Well I think I figured out a more permanent solution to this problem. From what I can determine, supplying 'bin_env: /usr/bin/pip3' actually leaves pip3 as the pip binary in the environment.
I've tried a few things, and it turns out that if we supply 'pip_bin: /usr/bin/python2' or 'pip_bin: /usr/bin/python3' that salt-call will automatically provide the '-m pip' to the bin and execute pip in the proper context, without putting a specific pip as the 'go-to' pip.

Also, when we choose '/usr/bin/pip' we're assuming that someone hasn't tried to avoid the python2/python3 problem by linking the pip3 binary to /usr/bin/pip so they can run tools/scripts/installs without specifying 3.

I got the idea from here: https://github.com/saltstack/salt/issues/48793
I tried specifying the pip_bin as the specific python version it needs, and so far so good.

I've got a DigitalOcean and AWS machine as well in the cloud, I'll try there as well and see what happens. I'm also going to try locally changing all the '/usr/bin/pip's to '/usr/bin/pythonX'

I'll keep you posted.
________________________________
From: Erik Kristensen <notifications@github.com>
Sent: Wednesday, May 20, 2020 5:52 PM
To: teamdfir/sift-saltstack <sift-saltstack@noreply.github.com>
Cc: Corey Forman <corey@digitalsleuth.ca>; Author <author@noreply.github.com>
Subject: Re: [teamdfir/sift-saltstack] Modify pip in appcompatprocessor.sls, update pdftk (#28)


Yeah I can. This is the problem. There's some weird case that I haven't been able to figure out yet where python3 takes precedence. This python2 vs python3 is becoming an issue. We maybe have to simply exclude ALL tools that cannot work in python3 until authors decide to update for python3.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/teamdfir/sift-saltstack/pull/28#issuecomment-631747011>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AO7OHXXZY6WZ5SRDDHZYHQ3RSRGKDANCNFSM4NDLQMJA>.

ekristen commented:
@digitalsleuth can you break this up into appcompat and pdftk. I want to get pdftk into master ASAP. I still can't verify this actually works. I'm attempting another install a different way.

digitalsleuth commented:
Yes, give me about 10 minutes.

digitalsleuth commented:
New PR submitted for pdftk only. You can close this if you want and we can revisit appcompat separately.

ekristen commented:
@digitalsleuth apologies but this is busted now with the inclusion of 20.04 support, can you take a look and let me know if any of these changes are still valid and or needed?

If you have an actual fix for appcompatprocessor on 20.04, let me know, because at the moment I cannot get it to install.

digitalsleuth commented:
After getting your message I decided to try installing it in Ubuntu 20.04 and did manage to get it to install, with a few tweaks here and there. 
- Python2.7 and get-pip.py are required (which you already have as sls in packages)
- python-dev-is-python2 is required for python-config
- build-essential, git, wget/curl to retrieve the libregf release
- download and unpack the latest libregf build [here](https://github.com/libyal/libregf/releases/download/20200805/libregf-alpha-20200805.tar.gz)
- enter the libregf directory and run 
```
./configure --enable-python && make
make install
python2.7 setup.py build
python2.7 setup.py install
pip2 install git+https://github.com/mbevilacqua/appcompatprocessor
ln -s /usr/local/lib/libregf.so.1 /usr/lib/libregf.so.1
```

Using this I managed to get it to install and run, but I have yet to test every option.
Since this seems like a lot of work to get this working, especially in a python2.7 environment, this can be closed unless it's useful.

ekristen commented:
We can’t do the Python dev is python2. I guess we are going to have to use docker 

Sent from my iPhone

> On Sep 19, 2020, at 20:01, Digital Sleuth <notifications@github.com> wrote:
> 
> ﻿
> After getting your message I decided to try installing it in Ubuntu 20.04 and did manage to get it to install, with a few tweaks here and there.
> 
> Python2.7 and get-pip.py are required (which you already have as sls in packages)
> python-dev-is-python2 is required for python-config
> build-essential, git, wget/curl to retrieve the libregf release
> download and unpack the latest libregf build here
> enter the libregf directory and run
> ./configure --enable-python && make
> make install
> python2.7 setup.py build
> python2.7 setup.py install
> pip2 install git+https://github.com/mbevilacqua/appcompatprocessor
> ln -s /usr/local/lib/libregf.so.1 /usr/lib/libregf.so.1
> Using this I managed to get it to install and run, but I have yet to test every option.
> Since this seems like a lot of work to get this working, especially in a python2.7 environment, this can be closed unless it's useful.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

digitalsleuth commented:
Makes sense. I just threw one together for you and tested it. Works well. Uploaded to hub.docker.com as well if you're interested.
https://github.com/digitalsleuth/appcompat-docker

digitalsleuth commented:
This pull request is being closed for the time being and can be re-opened at a later date if necessary.
